File: 프로그래머스/LV2/코딩테스트 연습/연습문제/혼자 놀기의 달인/혼자 놀기의 달인.py
def go(cards, idx, open, cnt):
    ret = 0
    if open[idx]: return cnt

    open[idx] = True
    ret = go(cards, cards[idx]-1, open, cnt + 1)

    return ret

def solution(cards):
    result = []
    open = [False] * len(cards)
    for i in range(len(cards)):
        if open[i]: continue
        result.append(go(cards, i, open, 0))

    result.sort(reverse=True)
    if len(result) == 1:
        return 0

    return result[0]*result[1]

File: 프로그래머스/LV2/코딩테스트 연습/연습문제/혼자 놀기의 달인/혼자 놀기의 달인.cpp
#include <string>
#include <vector>
#include <algorithm>
using namespace std;

int go(vector<int>& cards, int idx, vector<bool>& open, int cnt) {
    int ret = 0;
    if (open[idx])return cnt;

    open[idx] = true;
    ret = go(cards, cards[idx] - 1, open, cnt+1);

    return ret;
}

int solution(vector<int> cards) {
    int answer = 0;
    vector<int> result;
    vector<bool> open(cards.size(), false);

    for (int i = 0; i < cards.size(); i++) {
        if (open[i])continue;
        result.push_back(go(cards, i, open, 0));
    }

    sort(result.begin(), result.end());

    if (result.size() == 1)
        return 0;

    return result[0] * result[1];
}